### DISCUSSION TITLE
In exported site, next/navigation Router completely refreshes browser on router.push

### DISCUSSION THREAD

**Discussion by tcoz** [vercel/next.js#52464] [App Router]
This one is killing me, it has to be a simple fix but I can't figure it out.  I have looked everywhere and all I find is references to the older next/router using {shallow} (which doesn't work, you end up with "Router not Mounted" errors). 

Note I am using next/navigation { useRouter },  so no pages directly, no { shallow }. 

I have a simple app, a top-level "app" and two subfolders, books and houses:

app (with layout and page)
- books (with layout and page)
- houses (with layout and page)

In the app layout, I have a `<Header />`:

```
<html>
...
<body>
<Header />
{children}
</body>
</html>
```

In the Header, there is a menu. When you click a menu item (Books or Houses), it uses route.push("/books"). 

Note that I can not use Link. I'm using a PrimeReact menu, which works perfectly everywhere (except here). Each menu item takes an object from an "Items" collection, and each Item has a "command" on it you can use to run an operation. In this case, I need it to router.push("/books"). 

When I run the app "normally" (non-exported), it's fine. I can see the top-level state is retained (the theme the user selected, etc). The API calls for a given page aren't re-run (I'm using SWR, so the cache is being respected). 

When I export the app, every time the user navigates to another view (books or menu), the entire browser refreshes. All the API calls to populate the view are made again, and anything set into a Context (like the theme selection) is erased. 

I've tried everything, I can't figure this one out. In CRA it's absurdly easy, just use React Router and off you go.

Again, please note that this happens when I export the app. In the exported folder, I see all the separate pages based on the routes. But each page seems to be a complete version of the app, and the browser seems to just naively load the "new" page as if I was completely redirecting the site. 

Really stuck on this one.

**Comment by icyJoseph** [comment]
Hi,

I can't really reproduce this, I am using `export` mode as well, and a simple Header component with two buttons that call router push to `/about` or `/`, and I can do SPA transitions ~navigate~ without issue.

**Comment by tcoz** [comment]
The issue isn't that I can't navigate. The routes do load the correct pages. 

The issue is that the page refreshes when you navigate. I mean, the browser reloads. This means that every time I use the menu to select "home" or "books", all state is lost. It's as if the user hit the refresh button (though the target page does load). 

It can't work this way. If the user selects a Theme, which sets the info into a context, then navigates to another page and that context gets refreshed, it will switch back to the default theme, and so on. 

I'm going to set up a naive example and post it to GitHub. Hopefully I will solve my problem as I work through that example, but otherwise, if this is the behavior of an exported site for NextJS, the technology is largely useless to me. Figuring that can't be the case, I'll go with the assumption that I'm doing something wrong, though I appear to be following the documentation.

**Reply by icyJoseph** [reply]
I know what you mean. Mine doesn't unload the page, and any state I hold on the Header persists. That is to say, that SPA transitions work fine.

**Comment by tcoz** [comment]
Can you somehow post that code so I can see what you're doing? 

Also, this is using the next/navigation useRouter hook?

I'm following the documentation closely, but I did port this site from a CRA app. It seems pretty straightforward, but something else is going on. With an example and working through my dummy repo hopefully I can see where I'm going wrong.

**Reply by icyJoseph** [reply]
This is the Header:

```tsx
"use client";
import { useRouter } from "next/navigation";
import { useState } from "react";

export const Header = () => {
  const router = useRouter();
  const [count, setCount] = useState(0);
  return (
    <nav>
      <p>{count}</p>
      <button onClick={() => setCount((x) => x + 1)}>+1</button>
      <ul>
        <li>
          <button
            onClick={() => {
              router.push("/about");
            }}
          >
            Nav to About
          </button>
        </li>
        <li>
          <button
            onClick={() => {
              router.push("/");
            }}
          >
            Nav to Home
          </button>
        </li>
      </ul>
    </nav>
  );
};

```

And of course mate. Be mindful that I am genuinely after the root reason for your problem.

I wonder if there's somehow an anchor tag getting triggered on your setup, and nobody is calling `preventDefault` on it... but also weird that it happens only on exporting. 

I've seen an iteration of this bug where the link ends up being cut short of the last 4 characters... but this doesn't look like your issue.

**Comment by tcoz** [comment]
I'll have my repo done probably in an hour, I'll post my finding here. I'm going to use the same menu component (PrimeReact menu), and the same setup for state (a context) and switching themes. Everything else in the app is api calls and display, which shouldn't have anything to do with this (they are actions and elements of the sub-pages).

**Comment by tcoz** [comment]
I'm up and running, but still have the issue. I might have been naive in omitting a detail, on the surface it wouldn't seem to make a difference. 

I created a new NextJS app using npx create-next-app@latest. Then, copied the code into the project. Like, literally just deleted everything in "app", and dragged all my code into it. I cut and paste all the configs and the top-level layout and page files, installed my packages (SWR, animate.css, primereact, tailwind) made sure the next.config was set to output: 'export', and ran the build, went to the out directory, ran npx http-server, and lo and behold, everything works perfectly. 

So zero code changes. Literally straight drag 'n drops, cut 'n pastes. 

BUT...the old code was run in an Nx Workspace set to the "next" type. So it was a Next project in an Nx workspace. Otherwise, the Next project appears identical. Same package versions, everything. Instead of "out" though it writes to "dist", but otherwise there appears to be no difference. 

With one exception; the code runs properly in the stand-alone app. The export in the Nx Workspace has the issue. 

I'm mystified, but at this point am going to let the Nx Workspace thing go. Maybe I'll try rebuilding it and see what happens. 

No idea. The code is identical, the configurations are straightforward, and the output after running "build" is identical (though I haven't actually run diffs on the file contents).  

![Screen Shot 2023-07-10 at 9 50 31 AM](https://github.com/vercel/next.js/assets/1213203/25212f1f-29c7-4b43-a62a-1edc39534aae)

**Reply by icyJoseph** [reply]
Oh well, was kind of expecting something external to be the problem, but yeah, perhaps try to read in the `nx` docs. Also a good lesson on communicating, to focus on technical aspects, rather making subtle commentary on the framework.

**Comment by tcoz** [comment]
Went through them carefully (I've been working with Nx for a while, but never with a Next app). I'll have to do a teardown/rebuild of the workspace, repo it, and if I still see the issue, will forward it along. 

To the best of my knowledge I was very clear on the tech, and made no subtle commentary on the framework. Perhaps you could be more specific. I've reread the entire thread, perhaps consider if you are misconstruing intent, which is not hard to do in forum text, email, etc.
